# Movie Recommendation System: Model-Based Collaborative Filtering Model (Matrix Factorization)

**Background**:

An accurate, personalized recommendation system can improve business and sales and build customer satisfaction.


**Types of Recommendation Engines**:
1. **Popularity Model (see "popularity_movie_recommendation" notebook)**
2. **Recommendation algorithms**
    - **Content-based filtering (see "content_based_movie_recommendation" notebook)**
    - **Memory-based collaborative filtering (see "memory_based_collaborative_filtering_movie_recommendation" notebook)**
    - **Model-based collaborative filtering**
        - **Uses parametric machine learning to find user ratings of unrated items**
        - Examples: PCA, SVD, Neural Nets, Matrix Factorization
        - Learns parameters using gradient descent or other optimization algorithms
        - Applies dimensionality reduction techniques to capture more signals from the big datasets
        - **Advantages**:
            - Dimensionality reduction handles sparse/missing data
            - Avoids computational issues with large numbers of predictors
            - Easier to visualize the data
            - Easier to data mine with fewer predictors
            - Uncover hidden correlations/features in the raw data
            - Remove redundant and noise features that are useless
            - Access easier data storage and processing
        - **Disadvantages**:
            - Inference is intracable because factors are hidden/latent
        - Types of model-based collaborative filtering algorithms:
            1. **Matrix Factorization Based**
                - Unsupervised 
                - Deals better with scalability and sparsity than memory-based CF
                - Learns the latent prefernces of users and latent attributse of items from known ratings to predict the unknown ratings through the dot product of the latent features of users and items
                - Restructures the user_item matrix into low-rank structure and represent the matrix by the multiplication of two low-rank matrices where the rows contain the latent vector
                - Multiply the low-rank matrices together to approximate the original matrix and fill in any missing entries
                1. **Singular Vector Decomposition (see "model_based_cf_matrix_factorization_movie_recommendation" notebook)**
                    - Algorithm that decomposes a matrix A into two unitary matrices and a diagonal matrix
                    - A: input data matrix (ie user's ratings)
                    - U: left singular vectors (user "features" matrix) that represents how much users like each feature
                    - D: diagonal matrix of singular values (weights/strengths of each concept)
                    - V^T: right singular vectors (movie "features" matrix) that represents how relevant each feature is to each movie
                    - Reduces the dimension of the dataset
                    - Gives low-rank approximation of user tastes and preferences
                    - Need to extract user preferences that can be determined by a small number of hidden factors called **embeddings**
                    - **Embeddings** are low dimensional hidden factors for items and users
                    - Python packages for implementation: surprise
                2. **Probabilitistic Matrix Factorization**
                    - Python packages for implementation: fastai
                3. **Non -ve Matrix Factorization**
                    - Python packages for implementation: surprise
            2. **Deep learning/neural network (see "model_based_cf_deep_learning_movie_recommendation_model" notebook)**
                
3. **Using a classifier to make recommendation**
    - Classifiers are parametric solutions that require some parameters of the user and item to be defined first
    - Pros:
        - Incorporates personalization
        - Works even if the user's past history is short or not available
    - Cons:
        - Features might not be availalbe or sufficient to create a good classifier
        - Making a good classifier will become exponentially difficult as the number of user and items grow
        
(https://medium.com/@james_aka_yale/the-4-recommendation-engines-that-can-predict-your-movie-tastes-bbec857b8223)

**Data**:

We will be using the online movie recommender service MovieLens' dataset collected from the MovieLens website. The datasets were collected over several periods of time.
Users were selected at random to be included in the data. All users have rated 20+ movies. No demographic information is included.

The data includes:
- 100K ratings (1-5) from 1000 users on 1700 movies
- Each user has rated 20+ movies
- Simple demographic information for the users, such as gender, age, occupation, zip, etc.
- Genre information of movies

(https://grouplens.org/datasets/movielens/10m/)

In [1]:
import pandas as pd
import numpy as np
import scipy as sc
import pickle

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('fivethirtyeight')

import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 500)

# Data

## Users Data

In [2]:
users = pd.read_pickle('users.pickle')
users.head()

user_id  age sex  occupation zip_code
0        1   24   M  technician    85711
1        2   53   F       other    94043
2        3   23   M      writer    32067
3        4   24   M  technician    43537
4        5   33   F       other    15213

## Ratings Data

In [3]:
ratings = pd.read_pickle('ratings.pickle')
ratings.head()

user_id  movie_id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

## Movies Data

In [4]:
movies = pd.read_pickle('movies.pickle')
movies.drop(movies.columns[2:24], axis=1, inplace=True)
movies.head()

movie_id        movie_title                      genres
0         1   Toy Story (1995)  animation|childrens|comedy
1         2   GoldenEye (1995)   action|adventure|thriller
2         3  Four Rooms (1995)                    thriller
3         4  Get Shorty (1995)         action|comedy|drama
4         5     Copycat (1995)        crime|drama|thriller

# Model-Based Collaborative Filtering Movie Recommendation Model: Matrix Factorization with SVD

(https://medium.com/@james_aka_yale/the-4-recommendation-engines-that-can-predict-your-movie-tastes-bbec857b8223)

**Let's count the number of unique users and movies.**

In [5]:
num_users = ratings['user_id'].unique().shape[0]
num_movies = ratings['movie_id'].unique().shape[0]

print 'Number of users:', num_users
print 'Number of movies:', num_movies

Number of users: 943
Number of movies: 1682


**We need to format the ratings matrix to be one row per user and one column per movie.**

In [6]:
ratings_pivot = ratings.pivot(index='user_id', columns='movie_id', values='rating').fillna(0)
ratings_pivot.head()

movie_id  1     2     3     4     5     6     7     8     9     10    11    \
user_id                                                                      
1          5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0   2.0   
2          4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0   0.0   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   4.0   
5          4.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

movie_id  12    13    14    15    16    17    18    19    20    21    22    \
user_id                                                                      
1          5.0   5.0   5.0   5.0   5.0   3.0   4.0   5.0   4.0   1.0   4.0   
2          0.0   4.0   4.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0   0.0   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
5          0.0   0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0   3.0   0.0   

movie_id  23    24    25    26    27    28    29    30    31    32    33    \
user_id                                                                      
1          4.0   3.0   4.0   3.0   2.0   4.0   1.0   3.0   3.0   5.0   4.0   
2          0.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
5          0.0   4.0   3.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0   

movie_id  34    35    36    37    38    39    40    41    42    43    44    \
user_id                                                                      
1          2.0   1.0   2.0   2.0   3.0   4.0   3.0   2.0   5.0   4.0   5.0   
2          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
5          0.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0   5.0   0.0   0.0   

movie_id  45    46    47    48    49    50    51    52    53    54    55    \
user_id                                                                      
1          5.0   4.0   4.0   5.0   3.0   5.0   4.0   4.0   3.0   3.0   5.0   
2          0.0   0.0   0.0   0.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4          0.0   0.0   0.0   0.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0   
5          0.0   0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0   

movie_id  56    57    58    59    60    61    62    63    64    65    66    \
user_id                                                                      
1          4.0   5.0   4.0   5.0   5.0   4.0   3.0   2.0   5.0   4.0   4.0   
2          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
5          0.0   0.0   0.0   0.0   0.0   0.0   4.0   1.0   0.0   0.0   1.0   

movie_id  67    68    69    70    71    72    73    74    75    76    77    \
user_id                                                                      
1          3.0   4.0   3.0   3.0   3.0   4.0   3.0   1.0   4.0   4.0   4.0   
2          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
5          0.0   0.0   1.0   4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

movie_id  78    79    80    81    82    83    84    85    86    87    88    \
user_id                                                                      
1          1.0 

**Next, we need to denormalize the data (normalize by each user's mean) and convert it from a dataframe to a numpy array.**

In [7]:
ratings_pivot_matrix = ratings_pivot.as_matrix()
user_ratings_mean = np.mean(ratings_pivot_matrix, axis=1)
ratings_normalized = ratings_pivot_matrix - user_ratings_mean.reshape(-1, 1)

**Let's check the sparsity of the ratings dataset. When there is a very sparse matrix with lots of dimensions, doing matrix factorization can restructure the user-item matrix into low-rank structure to better approximate the original matrix.**

In [8]:
sparsity = round(1 - len(ratings)/float(num_users * num_movies), 3)
print 'The sparsity level of MovieLens100K dataset is', sparsity * 100, '%.'

The sparsity level of MovieLens100K dataset is 93.7 %.


**We will use the Scipy function `svds` because it allows us to choose how many latent factors to use to approximate the original ratings matrix.**

In [9]:
from scipy.sparse.linalg import svds
U, diags, Vt = svds(ratings_normalized, k=50)

In [10]:
D = np.diag(diags)

**We can start making movie predictions for every user from the decomposed matrices. But first we need to add the user means back to get the actual star ratings prediction.**

In [11]:
all_user_pred_ratings = np.dot(np.dot(U, D), Vt) + user_ratings_mean.reshape(-1, 1)

In [12]:
preds = pd.DataFrame(all_user_pred_ratings, columns=ratings_pivot.columns)
preds.head()

movie_id      1         2         3         4         5         6     \
0         6.488436  2.959503  1.634987  3.024467  1.656526  1.659506   
1         2.347262  0.129689 -0.098917  0.328828  0.159517  0.481361   
2         0.291905 -0.263830 -0.151454 -0.179289  0.013462 -0.088309   
3         0.366410 -0.443535  0.041151 -0.007616  0.055373 -0.080352   
4         4.263488  1.937122  0.052529  1.049350  0.652765  0.002836   

movie_id      7         8         9         10        11        12    \
0         3.630469  0.240669  1.791518  3.347816  2.966128  5.873074   
1         0.213002  0.097908  1.892100  0.671000  0.357140  0.121637   
2        -0.057624  0.568764 -0.018506  0.280742  0.142404  0.506428   
3         0.299015 -0.010882 -0.160888 -0.118834  0.769274  0.086523   
4         1.730461  0.870584  0.341027  0.569055 -0.675459  0.068190   

movie_id      13        14        15        16        17        18    \
0         4.101069  3.389306  4.106865  2.365477  2.051866  0.874133   
1         1.410628  3.077856  2.234084  0.466284  0.200723  0.133164   
2        -0.263161 -0.770552 -0.233785 -0.124033  0.181566 -0.099044   
3        -0.066764 -0.406609  0.628166  0.000970  0.065264 -0.025545   
4        -0.322558  0.032664  0.026021  0.075774  0.992908  0.109995   

movie_id      19        20        21        22        23        24    \
0         3.430062  2.858839  0.867343  2.910597  2.212223  3.262437   
1         1.948106  1.025987  0.059210 -0.019519 -0.035814  0.477771   
2        -0.467579 -0.046501  0.167284  0.541315  0.494334 -0.443344   
3        -0.212796 -0.144018 -0.094250  0.684112  0.215035 -0.218843   
4         0.053157 -0.320141  1.438065 -0.669422  0.147151  2.172482   

movie_id      25        26        27        28        29        30    \
0         4.399117  0.821052  1.236004  3.681455  0.583801  1.966995   
1         2.685548  0.314334  0.081742 -0.425463  0.157504 -0.059219   
2        -0.715635  0.174654 -0.055005  0.236562 -0.184623 -0.110396   
3         0.225003  0.142014 -0.086578  0.035754 -0.047350 -0.220572   
4         2.462140 -0.361953  0.336169  2.293966  1.264276  0.335608   

movie_id      31        32        33        34        35        36    \
0         3.699791  2.086317  2.275488  0.107836  0.302267  0.098060   
1         0.167320 -0.039025 -0.373729  0.079431 -0.072124  0.035314   
2        -0.046730  0.022726 -0.094683  0.021683 -0.004416 -0.002740   
3        -0.294959 -0.118327  0.046766  0.078226  0.023054  0.023894   
4         0.162204  0.604379  0.559965  0.037597  0.114350 -0.108925   

movie_id      37        38        39        40        41        42    \
0         0.390855  1.287024  2.429299  0.923861  0.293912  3.334737   
1         0.013882  0.150273  0.197624 -0.022988 -0.157782 -0.179674   
2         0.014568 -0.117285 -0.036792  0.006796 -0.040351 -0.169377   
3        -0.000842 -0.087519 -0.084298 -0.059802  0.190934  0.125090   
4         0.247861  0.888398  0.979366  1.170652  0.228253  1.891088   

movie_id      43        44        45        46        47        48    \
0         1.046127  2.125516  3.148509  0.847907  3.611339  1.611321   
1         0.140533  0.269805  0.664097 -0.163135  0.091001 -0.256106   
2        -0.028555  0.125309 -0.210559  0.002554  0.085771  0.329270   
3         0.053552  0.254264 -0.184201  0.050386  0.169078 -0.372409   
4        -0.111786  0.096035 -0.171570  0.300850  2.175281 -0.378666   

movie_id      49        50        51        52        53        54    \
0         2.307993  5.608465  0.741138  1.144871  1.456105  2.223851   
1         0.100469  3.535139 -0.222576 -0.197033  0.075891  0.063312   
2         0.042207  1.108815 -0.062797  0.147241 -0.157542  0.010112   
3        -0.193322  2.149895  0.076869 -0.307635  0.074349  0.076376   
4         0.284130  5.108787 -0.269969  0.548943  1.233711  0.399009   

movie_id      55        56        57        58        59        60    \
0         3.684991  4.517366  1

**We now create a function taht returns the movies with the highest predicted ratings that the specified user hasn't rated already.**

In [13]:
def rec_movies(predictions, user_id, movies, original_ratings, num_rec):
    
    # Get and sort the user's predictions
    user_row_num = user_id - 1
    sorted_user_preds = preds.iloc[user_row_num].sort_values(ascending=False)
    
    # Get the user's data and merge in the movie info
    user_data = original_ratings[original_ratings['user_id']==user_id]
    user_full = user_data.merge(movies, how='left', on='movie_id').sort_values(['rating'], ascending=False)
    print 'User {0} has already rated {1} movies.'.format(user_id, user_full.shape[0])
    print 'Recommending highest {0} predicted rating movies not already rated.'.format(num_rec)
    
    # Recommend the highest predicted rating movies that the user hasn't seen yet
    recs = movies[~movies['movie_id'].isin(user_full['movie_id'])].merge(pd.DataFrame(sorted_user_preds).reset_index(), how='left', on='movie_id').rename(columns={user_row_num: 'Predictions'}).sort_values('Predictions', ascending=False).iloc[:num_rec, :-1]
    
    return user_full, recs

**Let's recommend 20 movies to user id 210.**

In [14]:
already_rated, predictions = rec_movies(preds, 210, movies, ratings, 20)

User 210 has already rated 132 movies.
Recommending highest 20 predicted rating movies not already rated.


**Here are the top 20 movies user 210 has rated.**

In [15]:
already_rated.head(20)

user_id  movie_id  rating  timestamp  \
53       210       152       5  887730676   
44       210       182       5  887736232   
29       210        73       5  891035837   
82       210       483       5  887736482   
34       210        50       5  887731014   
81       210       181       5  887731082   
125      210       168       5  887730342   
79       210       482       5  887736739   
124      210       302       5  890059415   
41       210       257       5  887730789   
42       210       211       5  887730297   
123      210       174       5  887736045   
122      210         1       5  887731052   
115      210       655       5  887730496   
48       210       187       5  887736017   
50       210       276       5  887731147   
120      210        98       5  887736429   
55       210       197       5  887736393   
70       210        56       5  887730264   
59       210       127       5  887731230   

                                movie_title  \
53                           Sleeper (1973)   
44                        GoodFellas (1990)   
29                          Maverick (1994)   
82                        Casablanca (1942)   
34                         Star Wars (1977)   
81                Return of the Jedi (1983)   
125  Monty Python and the Holy Grail (1974)   
79                  Some Like It Hot (1959)   
124                L.A. Confidential (1997)   
41                      Men in Black (1997)   
42                           M*A*S*H (1970)   
123          Raiders of the Lost Ark (1981)   
122                        Toy Story (1995)   
115                      Stand by Me (1986)   
48           Godfather: Part II, The (1974)   
50                 Leaving Las Vegas (1995)   
120        Silence of the Lambs, The (1991)   
55                     Graduate, The (1967)   
70                      Pulp Fiction (1994)   
59                    Godfather, The (1972)   

                                  genres  
53                         comedy|sci_fi  
44                           crime|drama  
29                 action|comedy|western  
82                     drama|romance|war  
34   action|adventure|romance|sci_fi|war  
81   action|adventure|romance|sci_fi|war  
125                               comedy  
79                          comedy|crime  
124     crime|film_noir|mystery|thriller  
41        action|adventure|comedy|sci_fi  
42                            comedy|war  
123                     action|adventure  
122           animation|childrens|comedy  
115               adventure|comedy|drama  
48                    action|crime|drama  
50                         drama|romance  
120                       drama|thriller  
55                         drama|romance  
70                           crime|drama  
59                    action|crime|drama

**Here are the top 20 movies that our model-based CF model using matrix factorization recommends.**

In [16]:
predictions

movie_id                                        movie_title  \
135       183                                       Alien (1979)   
142       194                                  Sting, The (1973)   
63         82                               Jurassic Park (1993)   
111       143                         Sound of Music, The (1965)   
112       144                                    Die Hard (1988)   
103       133                          Gone with the Wind (1939)   
404       504                            Bonnie and Clyde (1967)   
52         66                     While You Were Sleeping (1995)   
171       239                                    Sneakers (1992)   
91        118                                     Twister (1996)   
198       269                             Full Monty, The (1997)   
170       237                               Jerry Maguire (1996)   
379       474  Dr. Strangelove or: How I Learned to Stop Worr...   
143       196                          Dead Poets Society (1989)   
342       432                                    Fantasia (1940)   
74         95                                     Aladdin (1992)   
419       521                            Deer Hunter, The (1978)   
38         47                                     Ed Wood (1994)   
319       405                         Mission: Impossible (1996)   
164       228                Star Trek: The Wrath of Khan (1982)   

                                 genres  
135       action|horror|sci_fi|thriller  
142                        comedy|crime  
63              action|adventure|sci_fi  
111                             musical  
112                     action|thriller  
103                   drama|romance|war  
404                         crime|drama  
52                       comedy|romance  
171                  crime|drama|sci_fi  
91            action|adventure|thriller  
198                              comedy  
170                       drama|romance  
379                          sci_fi|war  
143                               drama  
342         animation|childrens|musical  
74   animation|childrens|comedy|musical  
419                           drama|war  
38                         comedy|drama  
319            action|adventure|mystery  
164             action|adventure|sci_fi

**Although we didn't actually use the genre of the movies as a feature, the truncated matrix factorization features "picked up" on the underlying tastes of the user, as seen by the genres of the recommended movies and those of the user's top rated movies.**

# Evaluation the Model-Based CF Model Using Matrix Factorization

**We will use the `suprise` library instead because it provides powerful ready-to-use prediction algorithms including SVD to evaluate its RMSE. The `surprise` library is a Python scikit building and analyzing recommender systems.**

In [20]:
from surprise import Reader, Dataset, SVD, evaluate

In [23]:
# Load reader library
reader = Reader()

# Load ratings dataset with Dataset library
data = Dataset.load_from_df(ratings[['user_id', 'movie_id', 'rating']], reader)

# Split the dataset for 5-fold evaluation
data.split(n_folds=5)

In [24]:
svd = SVD()
evaluate(svd, data, measures=['RMSE'])

Evaluating RMSE of algorithm SVD.

------------
Fold 1
RMSE: 0.9375
------------
Fold 2
RMSE: 0.9369
------------
Fold 3
RMSE: 0.9323
------------
Fold 4
RMSE: 0.9363
------------
Fold 5
RMSE: 0.9351
------------
------------
Mean RMSE: 0.9356
------------
------------


CaseInsensitiveDefaultDict(list,
                           {'rmse': [0.93754135160943108,
                             0.93685829553426336,
                             0.93228252758368224,
                             0.93629478980283276,
                             0.93509856910237676]})

**The mean RMSE is 0.9356, which is pretty good. Let's now train on the dataset and get the predictions.**

In [25]:
trainset = data.build_full_trainset()
svd.train(trainset)

**Let's use the sample user 210 again.**

In [26]:
ratings[ratings['user_id']==210]

user_id  movie_id  rating  timestamp
13         210        40       3  891035994
1018       210       204       5  887730676
1155       210        70       4  887730589
1354       210        97       5  887736454
1417       210       527       5  887736232
1435       210       274       5  887730676
1850       210       357       5  887736206
2081       210       161       5  887736393
2392       210       380       4  887736482
2626       210        58       4  887730177
3121       210       502       3  891035965
3371       210       135       5  887736352
3452       210        96       4  887736616
3508       210       631       5  887736796
4123       210       134       5  887736070
4536       210       219       3  887808581
4657       210       255       4  887730842
4810       210       465       4  887737131
5353       210       722       4  891036021
5811       210       160       4  887737210
5923       210       501       4  887736998
6225       210       404       5  887736739
6961       210       216       4  887737603
7414       210       447       5  887737631
7829       210       202       5  887737338
7869       210        49       3  891036116
8212       210       186       4  887730532
8488       210       200       5  887737040
9399       210       176       4  887735960
9441       210        73       5  891035837
9847       210        25       4  887730407
9976       210       403       4  887736322
10091      210       484       4  887736070
10304      210       238       3  891036021
10357      210        50       5  887731014
10781      210       735       4  887737338
10973      210        44       3  887737710
11063      210       243       2  887734998
11112      210        65       4  887731305
11214      210       188       3  887737171
11688      210       185       4  887736232
11972      210       257       5  887730789
12504      210       211       5  887730297
12509      210       679       3  887808619
13106      210       182       5  887736232
13530      210       662       2  887730221
13536      210       692       4  887736796
13739      210       419       4  887737678
14423      210       187       5  887736017
15076      210       523       4  887730472
15101      210       276       5  887731147
15513      210       751       4  890059441
16152      210       953       3  887737488
17019      210       152       5  887730676
17043      210       864       3  887730842
17244      210       197       5  887736393
17245      210       300       4  887730066
17667      210        88       4  887737603
18743      210       179       3  887736429
19116      210       127       5  887731230
19631      210       210       5  887730532
19824      210       420       4  887737487
20136      210       926       2  887730909
20484      210       411       3  887730931
22166      210       154       4  887730341
22848      210      1012       4  887730789
23181      210       629       3  891035928
23248      210       172       5  887736261
23255      210        23       5  887730102
24136      210        94       4  891036181
24703      210        56       5  887730264
24737      210       180       4  887735872
24807      210       195       4  887736429
25484      210       517       4  887730342
25525      210       568       4  887735960
25695      210       327       4  887735288
26173      210       435       4  887730407
26795      210       290       4  887730813
27062      210        69       4  887736482
27395      210       482       5  887736739
27942      210       821       3  887730532
29375      210       181       5  887731082
29770      210       483       5  887736482
29919      210       208       5  887730443
32274      210       657       4  887736429
32746      210       301       4  887731435
32784      210         4       4  887730443
33958      210       956       3  887736900
35387      210       423       5  887737338
35421      210        99       4  887736937
359

**Let use SVD to predict the rating that user 210 will give to a random movie, ie movie id 835.**

In [27]:
svd.predict(210, 835)

Prediction(uid=210, iid=835, r_ui=None, est=4.2458539630091527, details={u'was_impossible': False})

**The estimated rating is 4.246. The recommender system works purely on the basis of an assigned movie ID and tries to predict ratings based on how the other users have predicted the movie.**

**Although the method of using matrix factorization (low-rank approximation) can capture the underlying features driving the raw data to scale better to large datasets and make better recommendations based on user tastes, we lose some meaningful signals. There is an interpretability problem as singular vector specifies a linear combination of all input columns or rows and there is a lack of sparsity when the singular vectors are quite dense; hence, the SVD method is limited to linear projections.**